### PR TITLE
Add support for X_CSI_SYSTEM_FILES variable to upload storage driver files

### PR DESCRIPTION
**Pull request by kirankt**
The X_CSI_SYSTEM_FILES env variable will be a tar file which will extract *any* required configuration files that the storage driver might require for startup. This tar file must be created with care by a privileged and trusted user. The purpose of using the X_CSI_SYSTEM_FILES is to make the ember CSI driver more generic in its configuration of its backend drivers.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -114,11 +114,23 @@ The CSI driver is configured via environmental variables, any value that doesn't
...
 | `X_CSI_ABORT_DUPLICATES`   | all        | If we want to abort or queue (default) duplicated requests.   | false                                                                                                    | true                                                                                                                                                                                                                                    |
 
 The only role that has been tested at the moment is the default one, where Controller and Node servicer are executed in the same service (`CSI_MODE=all`), and other modes are expected to have issues at the moment.
 
+The X_CSI_SYSTEM_FILES variable should point to a tar/tar.gz/tar.bz2 file, whose contents will be extracted into '/' of the Ember CSI driver container. An operator/administrator with privilege access must create the archive before starting the pods. Care must be taken not to overwrite the container files that may be critical to its startup.
```
Comment (Akrog) [inline README.md]: nit: Mention that when we are using containers this is the path within the container

File: ember_csi/ember_csi.py [hunk 2]
```diff
@@ -1026,6 +1027,25 @@ def _load_json_config(name, default=None):
         print('Invalid JSON data for %s' % name)
         exit(1)
 
+def copy_system_files():
+    # Check if the archive members are files only
+    # and not devices, etc
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: nit: This whole comment can go in a single line

File: ember_csi/ember_csi.py [hunk 3]
```diff
@@ -1026,6 +1027,25 @@ def _load_json_config(name, default=None):
...
+    def check_files(members):
+        for tarinfo in members:
+            if tarinfo.isfile() :
+                yield tarinfo
+
+    archive = os.environ.get('X_CSI_SYSTEM_FILES')
+    if archive is not None :
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: Better use: `if archive:` to ensure we also cover empty string.

File: ember_csi/ember_csi.py [hunk 4]
```diff
@@ -1026,6 +1027,25 @@ def _load_json_config(name, default=None):
         print('Invalid JSON data for %s' % name)
         exit(1)
 
+def copy_system_files():
+    # Check if the archive members are files only
+    # and not devices, etc
+    def check_files(members):
+        for tarinfo in members:
+            if tarinfo.isfile() :
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: We should not limit to files, as that would only support writing files to already existing directories, and we may want to create files in non-existent directories.

File: ember_csi/ember_csi.py [hunk 5]
```diff
@@ -1026,6 +1027,25 @@ def _load_json_config(name, default=None):
...
+                yield tarinfo
+
+    archive = os.environ.get('X_CSI_SYSTEM_FILES')
+    if archive is not None :
+        try:
+            with tarfile.open(archive, 'r') as t:
+                t.extractall('/', members=check_files(t))
+		t.close()
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: When using a context manager (`with`) we don't do the manual call to `close`

File: ember_csi/ember_csi.py [hunk 6]
```diff
@@ -1026,6 +1027,25 @@ def _load_json_config(name, default=None):
...
+        except Exception as exc:
+            sys.stderr.write('Error expanding file %s %s\n' % (archive, exc))
+            exit(2)
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: Maybe exit with error number 4

File: README.md (hunk 1 above)
Comment (kirankt) [reply README.md]: It does say: "extracted into ember csi driver container". Anyway, I've tweaked it a little.

File: ember_csi/ember_csi.py (hunk 2 above)
Comment (kirankt) [reply ember_csi/ember_csi.py]: Done

File: ember_csi/ember_csi.py (hunk 3 above)
Comment (kirankt) [reply ember_csi/ember_csi.py]: Makes sense. Thanks.

File: README.md (hunk 1 above)
Comment (kirankt) [reply README.md]: Updated README.

File: ember_csi/ember_csi.py (hunk 5 above)
Comment (kirankt) [reply ember_csi/ember_csi.py]: Ack.

File: README.md [hunk 7]
```diff
@@ -114,11 +114,23 @@ The CSI driver is configured via environmental variables, any value that doesn't
...
+The X_CSI_SYSTEM_FILES variable should point to a tar/tar.gz/tar.bz2 file accessible in the Ember CSI driver's filesystem. The contents of the archive will be extracted into '/'. A trusted user such as an operator/administrator with privileged access must create the archive before starting the driver.
+
+e.g.
+
+```
+$ tar cvf ceph-files.tar /etc/ceph/ceph.conf /etc/ceph/ceph.client.cinder.keyring
+tar: Removing leading `/' from member names
+/etc/ceph/ceph.conf
+/etc/ceph/ceph.client.cinder.keyring
+$ export X_CSI_SYSTEM_FILES=ceph-files.tar
```
Comment (Akrog) [inline README.md]: This should be the full path to the file...
```export X_CSI_SYSTEM_FILES=`pwd`/ceph-files.tar```

File: ember_csi/ember_csi.py [hunk 8]
```diff
@@ -1026,6 +1027,26 @@ def _load_json_config(name, default=None):
...
         exit(1)
 
+def copy_system_files():
+    # Minimal check of the archive for files/dirs only and not devices, etc
+    def check_files(members):
+        for tarinfo in members:
+            if tarinfo.isdev() :
+                sys.stderr.write("Skipping %s\n" % tarinfo.name)
+            else :
+                yield tarinfo
```
Comment (Akrog) [inline ember_csi/ember_csi.py]: We should yield after logging it

File: README.md (hunk 7 above)
Comment (kirankt) [reply README.md]: Done.

File: ember_csi/ember_csi.py (hunk 8 above)
Comment (kirankt) [reply ember_csi/ember_csi.py]: Done.

File: README.md [hunk 9]
```diff
@@ -129,7 +129,7 @@ $ tar cvf ceph-files.tar /etc/ceph/ceph.conf /etc/ceph/ceph.client.cinder.keyrin
 tar: Removing leading `/' from member names
 /etc/ceph/ceph.conf
 /etc/ceph/ceph.client.cinder.keyring
-$ export X_CSI_SYSTEM_FILES=ceph-files.tar
+$ export `pwd`/X_CSI_SYSTEM_FILES=ceph-files.tar
```
Comment (Akrog) [inline README.md]: pwd must be on the assignment side:

export X_CSI_SYSTEM_FILES=`pwd`/ceph-files.tar

Comment (Akrog) [reply README.md]: I just noticed this removed the ticks...  
```
X_CSI_SYSTEM_FILES=`pwd`/ceph-files.tar
```

**Pull request merged**